go-controller/pkg/libovsdbops/db_object_ids.go

```diff
@@ -0,0 +1,283 @@
+//
+//  dbIndex := NewDbObjectIDs(AddressSetEgressFirewallDNS, "DefaultController",
+//		map[ExternalIDKey]string{
+//			ObjectNameKey: "dns.name",
+// 			AddressSetIPFamilyKey: "ipv4"
+//	})
+// uses AddressSetEgressFirewallDNS = newObjectIDsType(addressSet, EgressFirewallDNSOwnerType, []ExternalIDKey{
+//		// dnsName
+//		ObjectNameKey,
+//		AddressSetIPFamilyKey,
+//  })
+//
+// its dbIndex will be mapped to the following ExternalIDs
+//  {
+//		"k8s.ovn.org/owner-controller": "DefaultController"
+//		"k8s.ovn.org/owner-type": "EgressFirewallDNS" (value of EgressFirewallDNSOwnerType)
+//		"k8s.ovn.org/name": "dns.name"
+// 		"k8s.ovn.org/ipFamily": "ipv4"
+//		"k8s.ovn.org/id": "DefaultController:EgressFirewallDNS:dns.name:ipv4"
+//  }
+type DbObjectIDs struct {
+	idsType ObjectIDsType
+	// ownerControllerName specifies which controller owns the object.
+	// Controller should only change objects it owns, make sure to always set this field.
+	ownerControllerName string
+	// objectIDs store values for keys required by given ObjectIDsType, and may be different for different ObjectIDsType.
+	// These ids should uniquely identify db object with the same ownerControllerName and OwnerTypeKey.
+	objectIDs map[ExternalIDKey]string
+}
+
+// NewDbObjectIDs is used to construct DbObjectIDs, idsType and controller are always required,
+// objectIds may be empty, or half-filled for predicate search.
+// objectIds keys that are not used by given idsType will not be used.
+func NewDbObjectIDs(idsType ObjectIDsType, controller string, objectIds map[ExternalIDKey]string) *DbObjectIDs {
+	if controller == "" {
+		panic("NewDbObjectIDs failed: controller should not be empty")
+	}
+	externalIDKeys := idsType.GetExternalIDKeys()
+	if externalIDKeys == nil {
+		// can only happen if ObjectIDsType{} is passed
+		panic(fmt.Sprintf("NewDbObjectIDs failed: ObjectIDsType %v should not be empty", idsType))
+	}
+	// only use values for keys from idsType
+	filteredObjectIDs := map[ExternalIDKey]string{}
+	for _, externalIDKey := range externalIDKeys {
+		if value, ok := objectIds[externalIDKey]; ok {
+			filteredObjectIDs[externalIDKey] = value
+		}
+	}
+	objectIDs := &DbObjectIDs{
+		idsType:             idsType,
+		ownerControllerName: controller,
+		objectIDs:           filteredObjectIDs,
+	}
+	return objectIDs
+}
+
+func (objectIDs *DbObjectIDs) NewDbObjectWithExtraIDs(extraObjectIds map[ExternalIDKey]string) *DbObjectIDs {
+	ids := deepcopyMap(objectIDs.objectIDs)
+	for key, value := range extraObjectIds {
+		ids[key] = value
+	}
+	return NewDbObjectIDs(objectIDs.idsType, objectIDs.ownerControllerName, ids)
+}
+
+func (objectIDs *DbObjectIDs) NewDbObjectWithoutIDs(idsToDelete ...ExternalIDKey) *DbObjectIDs {
+	ids := deepcopyMap(objectIDs.objectIDs)
+	for _, keyToDel := range idsToDelete {
+		delete(ids, keyToDel)
+	}
+	return NewDbObjectIDs(objectIDs.idsType, objectIDs.ownerControllerName, ids)
+}
```

Comment:
yeah that's because I have just added copying on `NewDbObjectIDs`, good catch

Reply:
so based on the usual usage to reduce the number of copies, I updated `NewDbObjectIDs` to use given map and delete extra keys form it directly (because most of the times that map will created only to call `NewDbObjectIDs` and won't contain any extra ids)
that means I need to leave `deepcopyMap`, but I renamed the functions as you suggested, hopefully it is not too implicit

Reply:
Now you have an extra loop. LOL don't mind me, its my past of system engineer.

If you keep `deepCopyMap`, why wouldn't you though:
```
func (objectIDs *DbObjectIDs) AddIDs(extraObjectIds map[ExternalIDKey]string) *DbObjectIDs {
	ids := deepcopyMap(objectIDs.objectIDs)
	for key, value := range extraObjectIds {
                if objectIDs.idsType.HasKey(key) {
			ids[key] = value
		}
	}
	return &DbObjectIDs{objectIDs.idsType, objectIDs.ownerControllerName, ids}
}

func (objectIDs *DbObjectIDs) RemoveIDs(idsToDelete ...ExternalIDKey) *DbObjectIDs {
	ids := deepcopyMap(objectIDs.objectIDs)
	for _, keyToDel := range idsToDelete {
		delete(ids, keyToDel)
	}
	return &DbObjectIDs{objectIDs.idsType, objectIDs.ownerControllerName, ids}
}
```

Reply:
initially I wanted to keep that `if idsType.HasKey(key)` logic in `NewDbObjectIDs` only, now it will be spread across 2 functions. But it is better for performance, as you said, so let's go that way